--- capi4hylafax-01.03.00/src/capi/CCStruct.h	2016-10-16 21:31:43.967938981 +0200
+++ capi4hylafax_01.03.00_vb/src/capi/CCStruct.h	2016-10-15 23:27:44.210169153 +0200
@@ -141,7 +141,7 @@
 class CCStruct {
 public:
     CCStruct (tCapiStructType datatype = cxt_RawData);
-    CCStruct (const void *data, tUInt len, tCapiStructType datatype = cxt_RawData);
+    CCStruct (const void *data, tUInt len, const tCapiStructType datatype = cxt_RawData);
     virtual ~CCStruct (void);
     void Reset (void);
 
@@ -170,21 +170,21 @@
     tCapiStructType GetClassType (void);
 
 protected:
-    CCStruct (void *data, tFormatChar *format);
-    CCStruct (tCapiStructType type, void *data, tUInt len, tFormatChar *format = cxf_CONST);
-    CCStruct (tCapiStructType type, void *data, tFormatChar *format);
+    CCStruct (void *data, const tFormatChar *format);
+    CCStruct (tCapiStructType type, void *data, tUInt len, const tFormatChar *format = cxf_CONST);
+    CCStruct (tCapiStructType type, void *data, const tFormatChar *format);
 
 private:
     void   FillData  (tUByte **ppBuffer);
     tSInfo ParseData (tUByte **ppBuffer, tUInt Len);
     tBool  Resize    (tUInt maxlen);
 
-    tCapiStructType m_ClassType;
-    tCapiStructType m_DataType;
-    tFormatChar    *m_Format;
-    tUInt           m_Len;
-    tUInt           m_MaxLen;
-    tUByte         *m_pData;
+    tCapiStructType       m_ClassType;
+    tCapiStructType       m_DataType;
+    const tFormatChar    *m_Format;
+    tUInt                 m_Len;
+    tUInt                 m_MaxLen;
+    tUByte               *m_pData;
 };
 
 
@@ -219,7 +219,7 @@
     m_pData ((tUByte *)data) {
 }
 
-inline CCStruct::CCStruct (void *data, tFormatChar *format)
+inline CCStruct::CCStruct (void *data, const tFormatChar *format)
   : m_ClassType (cxt_StructData),
     m_DataType (cxt_StructData),
     m_Format (format),
@@ -228,7 +228,7 @@
     m_pData ((tUByte *)data) {
 }
 
-inline CCStruct::CCStruct (tCapiStructType type, void *data, tUInt len, tFormatChar *format)
+inline CCStruct::CCStruct (tCapiStructType type, void *data, tUInt len, const tFormatChar *format)
   : m_ClassType (type),
     m_DataType (type),
     m_Format (format),
@@ -239,7 +239,7 @@
     // dassert (type != cxt_RawData);
 }
 
-inline CCStruct::CCStruct (tCapiStructType type, void *data, tFormatChar *format)
+inline CCStruct::CCStruct (tCapiStructType type, void *data, const tFormatChar *format)
   : m_ClassType (type),
     m_DataType (type),
     m_Format (format),
--- capi4hylafax-01.03.00/src/capi/CCSMesg.h	2016-10-16 21:31:43.967938981 +0200
+++ capi4hylafax_01.03.00_vb/src/capi/CCSMesg.h	2016-10-15 23:27:44.210169153 +0200
@@ -43,10 +43,10 @@
 
 class cm_messagebody : public CCStruct {
 protected:
-    cm_messagebody (tCapiStructType type, void *data, tUInt len, tFormatChar *format)
+    cm_messagebody (tCapiStructType type, void *data, tUInt len, const tFormatChar *format)
       : CCStruct (type, data, len, format) {}
 
-    cm_messagebody (tCapiStructType type, void *data, tFormatChar *format)
+    cm_messagebody (tCapiStructType type, void *data, const tFormatChar *format)
       : CCStruct (type, data, format) {}
 };
 
--- capi4hylafax-01.03.00/src/capi/CapiStat.cpp	2016-10-16 21:31:43.971938875 +0200
+++ capi4hylafax_01.03.00_vb/src/capi/CapiStat.cpp	2016-10-15 23:27:44.210169153 +0200
@@ -376,11 +376,21 @@
     if (m_pCurrentData && m_pCurrentData->pControllerData) {
         dassert (Controller > 0);
         dassert (Controller <= m_pCurrentData->CountControllers);
-        if (  (CheckOneFlag (m_pCurrentData->pControllerData[Controller - 1].B1ProtocolSupport, b1prot_ATM) == vTrue)
-           && (*(tUInt32 *)m_pCurrentData->pControllerData[Controller - 1].DSL_MACAddress != 0)) {
-            // seems to be a valid MAC Address
-            s_memcpy (pAddress, m_pCurrentData->pControllerData[Controller - 1].DSL_MACAddress, 6);
-            fret = vTrue;
+        if (  (CheckOneFlag (m_pCurrentData->pControllerData[Controller - 1].B1ProtocolSupport, b1prot_ATM) == vTrue)) {
+//           && (*(tUInt32 *)m_pCurrentData->pControllerData[Controller - 1].DSL_MACAddress != 0)) {
+            
+            bool all_mac_bytes_zero = true;
+            for (size_t mac_byte_idx=0; mac_byte_idx < sizeof(((ctIntCntrlData*)NULL)->DSL_MACAddress); ++mac_byte_idx) {
+                if (m_pCurrentData->pControllerData[Controller - 1].DSL_MACAddress[mac_byte_idx] != 0) {
+                    all_mac_bytes_zero = false;
+                }
+            }
+           
+            if (!all_mac_bytes_zero) {
+                // seems to be a valid MAC Address
+                s_memcpy (pAddress, m_pCurrentData->pControllerData[Controller - 1].DSL_MACAddress, 6);
+                fret = vTrue;
+            }
         }
     }
     m_Protect.EndRead();
--- capi4hylafax-01.03.00/src/capi/CCStruct.cpp	2016-10-16 21:31:43.971938875 +0200
+++ capi4hylafax_01.03.00_vb/src/capi/CCStruct.cpp	2016-10-15 23:27:44.210169153 +0200
@@ -74,7 +74,7 @@
 
 void CCStruct::Reset (void) {
     dhead ("CCStruct::Reset", DCON_CCStruct);
-    tFormatChar *format = m_Format;
+    const tFormatChar *format = (char*)m_Format;
     switch (*format) {
     case cxfv_NOLEN:
     case cxfv_MSGLEN:
@@ -220,7 +220,7 @@
     dparams ("%x,%x,l=%x", m_ClassType, m_DataType, m_Len);
     dassert (m_Format != 0);
     if (m_Len == 0) {
-        tFormatChar *format = m_Format;
+        const tFormatChar *format = m_Format;
         tUByte      *pSData = m_pData;
         switch (*format) {
         case cxfv_CONST:
@@ -385,7 +385,7 @@
 
 void CCStruct::ResetStructSize (tBool recursive) {
     dhead ("CCStruct::ResetStructSize", DCON_CCStruct);
-    tFormatChar *format = m_Format;
+    const tFormatChar *format = m_Format;
     switch (*format) {
     case cxfv_CONST:
         break;
@@ -664,7 +664,7 @@
     dassert (m_Len <= MAXVAL_tUShort);
     dassert ((m_pData != 0) || (m_Len == 0));
     dassert (m_Format != 0);
-    tFormatChar *format = m_Format;
+    const tFormatChar *format = m_Format;
     switch (*format) {
     case cxfv_NOLEN:
     case cxfv_MSGLEN:
@@ -851,7 +851,7 @@
     dassert (*ppBuffer != 0);
     dassert (m_Format != 0);
     dprintBuf (*ppBuffer, len);
-    tFormatChar *format = m_Format;
+    const tFormatChar *format = m_Format;
     switch (*format) {
     case cxfv_NOLEN:
     case cxfv_MSGLEN:
--- capi4hylafax-01.03.00/src/capi/DCMTrace.cpp	2016-10-16 21:31:43.971938875 +0200
+++ capi4hylafax_01.03.00_vb/src/capi/DCMTrace.cpp	2016-10-15 23:27:44.210169153 +0200
@@ -25,7 +25,7 @@
 \*===========================================================================*/
 
 static tUChar *DbgStrCapiCommand (c_byte Command, c_byte SubCommand) {
-    char *fret;
+    const char *fret;
     switch (SubCommand) {
     case CAPI_REQ:
         switch (Command) {
--- capi4hylafax-01.03.00/src/capi/CCSParam.h	2016-10-16 21:31:43.971938875 +0200
+++ capi4hylafax_01.03.00_vb/src/capi/CCSParam.h	2016-10-15 23:27:44.214169049 +0200
@@ -175,7 +175,7 @@
 
 class cp_b1configuration : public CCStruct {
 protected:
-    cp_b1configuration (tCapiStructType type, void *data, tFormatChar *format)
+    cp_b1configuration (tCapiStructType type, void *data, const tFormatChar *format)
       : CCStruct (type, data, format) {}
 };
 
@@ -262,7 +262,7 @@
 
 class cp_b3configuration : public CCStruct {
 protected:
-    cp_b3configuration (void *data, tFormatChar *format)
+    cp_b3configuration (void *data, const tFormatChar *format)
       : CCStruct (cxt_B1Config, data, format) {}
 };
 
@@ -410,7 +410,7 @@
 
 class cp_ncpi : public CCStruct {
 protected:
-    cp_ncpi (void *data, tFormatChar *format)
+    cp_ncpi (void *data, const tFormatChar *format)
       : CCStruct (cxt_Ncpi, data, format) {}
 };
 
--- capi4hylafax-01.03.00/src/capi/CapiStat.h	2016-10-16 21:31:43.971938875 +0200
+++ capi4hylafax_01.03.00_vb/src/capi/CapiStat.h	2016-10-15 23:27:44.214169049 +0200
@@ -104,6 +104,22 @@
         tUInt   DSL_Speed;
         tUByte  DSL_MACAddress [6];
         tUByte  DSL_Retries;
+    
+        ctIntCntrlData() : 
+           NumBChannels(0),
+           GlobalOptions(0),
+           B1ProtocolSupport(0),
+           B2ProtocolSupport(0),
+           B3ProtocolSupport(0),
+           SupplServices(0),
+           DSL_ModemState(0),
+           DSL_Speed(0),
+           DSL_Retries(0) {
+            for (unsigned int i=0; i<6; i++) {
+                DSL_MACAddress[i] = 0;
+            }        
+        }
+    
     };
     struct ctIntCapiData {
         c_info          CapiInstalled;
--- capi4hylafax-01.03.00/src/convert/BasicFax.h	2016-10-16 21:31:43.975938770 +0200
+++ capi4hylafax_01.03.00_vb/src/convert/BasicFax.h	2016-10-15 23:27:44.218168944 +0200
@@ -36,7 +36,7 @@
     virtual ~CBasicFaxConversion (void);
     void ConnectLog (cp_ncpi_all *pNCPI);
     void WriteXferLog (const char *cmd, tUInt commid, tUChar *jobtag, tUInt *pJobid, char *devname,
-                       CDynamicString *pUser, CDynamicString *pDestination, char *statustext);
+                       CDynamicString *pUser, CDynamicString *pDestination, const char *statustext);
 
     tSInfo OpenSend (CMultiString *pSendFileList, tUInt Resolution);
     void StopSend (void);
@@ -69,7 +69,7 @@
     time_t GetStartTime (void);
 
     virtual void PageEndReached (tBool IsLastPage) = 0;
-    virtual void WriteLog (tSInt /*priority*/, char * /*text*/, ...) {}
+    virtual void WriteLog (tSInt /*priority*/, const char * /*text*/, ...) {}
     virtual tUInt GetRecvFileMode (void) { return DEFAULT_FILE_MODE; }
     virtual CDynamicString *GetSpoolDir (void) { return 0; }
 
--- capi4hylafax-01.03.00/src/convert/BasicFax.cpp	2016-10-16 21:31:43.975938770 +0200
+++ capi4hylafax_01.03.00_vb/src/convert/BasicFax.cpp	2016-10-15 23:27:44.218168944 +0200
@@ -41,7 +41,7 @@
 
 void fmtTime (CDynamicString *pString, time_t t) {
     time_t v;
-    char  *fmt = "%u:%02u";
+    const char  *fmt = "%u:%02u";
     if (t < 0) {
         pString->Append ("0:00", 4);
     } else if (t > 99*60*60) {
@@ -164,7 +164,7 @@
 \*===========================================================================*/
 
 void CBasicFaxConversion::WriteXferLog (const char *cmd, tUInt commid, tUChar *jobtag, tUInt *pJobid, char *devname,
-                                        CDynamicString *pUser, CDynamicString *pDestination, char *statustext) {
+                                        CDynamicString *pUser, CDynamicString *pDestination, const char *statustext) {
     dhead ("CBasicFaxConversion::WriteXferLog", DCON_CBasicFaxConversion);
     dassert (cmd != 0);
     dassert (devname != 0);
@@ -174,7 +174,7 @@
         return;
     }
     char buf[80];
-    char *timestrf = "%D %H:%M";
+    const char *timestrf = "%D %H:%M";
     strftime (buf, sizeof (buf), timestrf, localtime (&m_StartTime));
     CDynamicString xferline (200);
     xferline.Print ("%s\t%s\t%09u\t%s", buf, cmd, commid, devname);
@@ -253,7 +253,7 @@
     m_StopTime            = m_StartTime;
     m_ConnectStartTime    = 0;
     m_BaudRate            = 0;
-    char *ModeText        = "SFF";
+    const char *ModeText        = "SFF";
     switch (m_format) {
     default:
         dassert (0);
--- capi4hylafax-01.03.00/src/faxrecv/faxrecv.h	2016-10-16 21:31:43.975938770 +0200
+++ capi4hylafax_01.03.00_vb/src/faxrecv/faxrecv.h	2016-10-15 23:27:44.218168944 +0200
@@ -53,7 +53,7 @@
 
     // virtual functions from CBasicFaxConversion
     void PageEndReached (tBool IsLastPage);
-    void WriteLog (tSInt priority, char *text, ...);
+    void WriteLog (tSInt priority, const char *text, ...);
     tUInt GetRecvFileMode (void);
     CDynamicString *GetSpoolDir (void);
 
--- capi4hylafax-01.03.00/src/faxrecv/recvmain.h	2016-10-16 21:31:43.975938770 +0200
+++ capi4hylafax_01.03.00_vb/src/faxrecv/recvmain.h	2016-10-15 23:27:44.218168944 +0200
@@ -33,8 +33,8 @@
     CFaxReceiveMain (void);
     ~CFaxReceiveMain (void);
 
-    void WriteLog (tSInt priority, char *text, ...);
-    void vWriteLog (tSInt priority, char *text, va_list params);
+    void WriteLog (tSInt priority, const char *text, ...);
+    void vWriteLog (tSInt priority, const char *text, va_list params);
     void GenerateHylafaxConfigFile (CDynamicString *pName, CDynamicString *pFaxNumber, CDynamicString *pIdentifier);
 
     tBool Config (char *CommandLineSpoolDir);
--- capi4hylafax-01.03.00/src/faxrecv/recvdev.h	2016-10-16 21:31:43.975938770 +0200
+++ capi4hylafax_01.03.00_vb/src/faxrecv/recvdev.h	2016-10-15 23:27:44.218168944 +0200
@@ -43,13 +43,13 @@
     tBool RecreateFIFO (void);
     tSInt GetFIFOHandle (void);
     void FIFOchanged (void);
-    tBool NotifyFIFO (tBool Receive, char Type, char *Data);
-    tBool NotifyFIFO (char Type);
+    tBool NotifyFIFO (tBool Receive, const char Type, const char *Data);
+    tBool NotifyFIFO (const char Type);
 
     // write functions
-    void WriteLog (tSInt priority, char *text, ...);
-    void vWriteLog (tSInt priority, char *text, va_list args);
-    void WriteStatus (char *text);
+    void WriteLog (tSInt priority, const char *text, ...);
+    void vWriteLog (tSInt priority, const char *text, va_list args);
+    void WriteStatus (const char *text);
 
     CDynamicString   DeviceName;
     CFaxReceiveMain *MainClass;
--- capi4hylafax-01.03.00/src/faxrecv/recvdev.cpp	2016-10-16 21:31:43.975938770 +0200
+++ capi4hylafax_01.03.00_vb/src/faxrecv/recvdev.cpp	2016-10-15 23:27:44.218168944 +0200
@@ -421,7 +421,7 @@
 /*===========================================================================*\
 \*===========================================================================*/
 
-tBool CFaxReceiveDevice::NotifyFIFO (tBool Receive, char Type, char *Data) {
+tBool CFaxReceiveDevice::NotifyFIFO (tBool Receive, const char Type, const char *Data) {
     dhead ("CFaxReceiveDevice::NotifyFIFO", DCON_CFaxReceive);
     tBool fret = vFalse;
     if (format != FaxFormat_Hylafax) {
@@ -455,10 +455,10 @@
         return vTrue;
     }
     tBool  recv = vFalse;
-    char  *text = 0;
+    const char  *text = 0;
     switch (Type) {
     case 'N':
-        text = (char *)FaxNumber.GetPointer();
+        text = (const char *)FaxNumber.GetPointer();
         break;
     case 'R':
         recv = vFalse;
@@ -478,7 +478,7 @@
 /*===========================================================================*\
 \*===========================================================================*/
 
-void CFaxReceiveDevice::WriteLog (tSInt priority, char *text, ...) {
+void CFaxReceiveDevice::WriteLog (tSInt priority, const char *text, ...) {
     dhead ("CFaxReceiveDevice::WriteLog", DCON_CFaxReceive);
     va_list params;
     va_start (params, text);
@@ -489,7 +489,7 @@
 /*===========================================================================*\
 \*===========================================================================*/
 
-void CFaxReceiveDevice::vWriteLog (tSInt priority, char *text, va_list args) {
+void CFaxReceiveDevice::vWriteLog (tSInt priority, const char *text, va_list args) {
     dhead ("CFaxReceiveDevice::WriteLog", DCON_CFaxReceive);
     MainClass->vWriteLog (priority, text, args);
 }
@@ -497,7 +497,7 @@
 /*===========================================================================*\
 \*===========================================================================*/
 
-void CFaxReceiveDevice::WriteStatus (char *text) {
+void CFaxReceiveDevice::WriteStatus (const char *text) {
     dhead ("CFaxReceiveDevice::WriteStatus", DCON_CFaxReceive);
     if (hStatusFile == 0) {
         return;
--- capi4hylafax-01.03.00/src/faxrecv/recvmain.cpp	2016-10-16 21:31:43.975938770 +0200
+++ capi4hylafax_01.03.00_vb/src/faxrecv/recvmain.cpp	2016-10-15 23:27:44.218168944 +0200
@@ -69,16 +69,16 @@
 
 CFaxReceiveMain::~CFaxReceiveMain (void) {
     dhead ("CFaxReceiveMain-Destructor", DCON_CFaxReceiveMain);
-    if (hLogFile > 0) {
+    if (hLogFile != NULL) {
         fclose (hLogFile);
-        hLogFile = 0;
+        hLogFile = NULL;
     }
 }
 
 /*===========================================================================*\
 \*===========================================================================*/
 
-void CFaxReceiveMain::WriteLog (tSInt priority, char *text, ...) {
+void CFaxReceiveMain::WriteLog (tSInt priority, const char *text, ...) {
     dhead ("CFaxReceiveMain::WriteLog", DCON_CFaxReceiveMain);
     va_list params;
     va_start (params, text);
@@ -89,9 +89,9 @@
 /*===========================================================================*\
 \*===========================================================================*/
 
-void CFaxReceiveMain::vWriteLog (tSInt priority, char *text, va_list params) {
+void CFaxReceiveMain::vWriteLog (tSInt priority, const char *text, va_list params) {
     dhead ("CFaxReceiveMain::vWriteLog", DCON_CFaxReceiveMain);
-    char   *statusText = 0;
+    const char   *statusText = NULL;
     char    timebuf[MAX_STRING_SIZE];
     va_list lparams;
     timeval tv;
@@ -323,7 +323,7 @@
                   "The directory is missing or inaccessible.\n", SpoolDir.GetPointer());
         return vFalse;
     }
-    char *ModeText = "Hylafax";
+    const char *ModeText = "Hylafax";
     switch (format) {
     case FaxFormat_TIFF:
         ModeText = "TIFF";
@@ -527,7 +527,7 @@
 
 tSInt CFaxReceiveMain::main (tSInt argc, char **argv) {
     dhead ("CFaxReceive::main", DCON_CFaxReceiveMain);
-    char *deviceName  = DEFAULT_DEVICE_NAME;
+    char deviceName[255]  = DEFAULT_DEVICE_NAME;
     char *cl_spooldir = 0;
     int   ch;
 
@@ -593,11 +593,13 @@
     }
     dprint ("%x,%x-", optind, argc);
     if (optind < argc) {
-        deviceName = strrchr (argv[optind], '/');
-        if (deviceName) {
-            deviceName++;
+        char * deviceName_in = strrchr (argv[optind], '/');
+        if (deviceName_in) {
+            deviceName_in++;
+            strncpy(deviceName, deviceName_in, sizeof(deviceName));
         } else {
-            deviceName = argv[optind];
+            //deviceName = argv[optind];
+            strncpy(deviceName, argv[optind], sizeof(deviceName));
         }
     }
     if (!configName) {
--- capi4hylafax-01.03.00/src/faxrecv/faxrecv.cpp	2016-10-16 21:31:43.975938770 +0200
+++ capi4hylafax_01.03.00_vb/src/faxrecv/faxrecv.cpp	2016-10-15 23:27:44.218168944 +0200
@@ -74,7 +74,7 @@
 CFaxReceive::~CFaxReceive (void) {
     dhead ("CFaxReceive - Destructor", DCON_CFaxReceive);
     CloseRecv();
-    if (hJobFile > 0) {
+    if (hJobFile != NULL) {
         fclose (hJobFile);
         hJobFile = 0;
     }
@@ -306,7 +306,7 @@
         if (FaxDevice->MainClass->FaxRcvdCmd.IsEmpty() == vFalse) {
             CDynamicString executeCommand;
             executeCommand.Set (&FaxDevice->MainClass->FaxRcvdCmd);
-            char *StateText = "";
+            const char *StateText = "";
 
             if (GetFormat() == FaxFormat_Hylafax) {
                 switch (GetRecvState()) {
@@ -390,10 +390,10 @@
         GetRecvFiles()->RemoveAndDeleteAll();
     }
     WriteLog (LOG_INFO, "Connection is droped with reason 0x%X (%s).\n", Reason, capi_info2str (Reason));
-    if (hJobFile > 0) {
+    if (hJobFile != NULL) {
         WriteLog (LOG_INFO, "SESSION END\n");
         fclose (hJobFile);
-        hJobFile = 0;
+        hJobFile = NULL;
     }
     jobNr       = 0;
     ONumber.RemoveAll();
@@ -430,7 +430,7 @@
 /*===========================================================================*\
 \*===========================================================================*/
 
-void CFaxReceive::WriteLog (tSInt priority, char *text, ...) {
+void CFaxReceive::WriteLog (tSInt priority, const char *text, ...) {
     dhead ("CFaxReceive::WriteLog", DCON_CFaxReceive);
     char    timebuf[MAX_STRING_SIZE];
     timeval tv;
--- capi4hylafax-01.03.00/src/faxsend/faxsend.h	2016-10-16 21:31:43.975938770 +0200
+++ capi4hylafax_01.03.00_vb/src/faxsend/faxsend.h	2016-10-15 23:27:44.222168839 +0200
@@ -38,7 +38,7 @@
     tSInfo Send (CDynamicString *Number, CMultiString *FaxFilesList, tUInt Resolution,
                  tUInt jobID, tUInt *pCommID, CDynamicString *pPollStr);
 
-    void WriteLog (tSInt priority, char *text, ...);
+    void WriteLog (tSInt priority, const char *text, ...);
 
     tUInt exitState;
     void CloseJobFile (void);
--- capi4hylafax-01.03.00/src/faxsend/sendmain.cpp	2016-10-16 21:31:43.975938770 +0200
+++ capi4hylafax_01.03.00_vb/src/faxsend/sendmain.cpp	2016-10-15 23:27:44.222168839 +0200
@@ -47,7 +47,7 @@
 /*===========================================================================*\
 \*===========================================================================*/
 
-static void WriteQFile (tChar *QFileName, tUInt commID, tUInt HylafaxExitState, CFaxSend *fs, char **statustext) {
+static void WriteQFile (tChar *QFileName, tUInt commID, tUInt HylafaxExitState, CFaxSend *fs, const char **statustext) {
     dhead ("WriteQFile", DCON_CFaxSend);
     dassert (QFileName != 0);
     CDynamicQueue textQ (60, sizeof (char *));
@@ -240,7 +240,7 @@
     tUInt             JobID         = 0;
     tUInt             commID        = 0;
     tUInt             HylaExitState = 1;
-    char             *statustext;
+    const char             *statustext;
     CConfigFileParser FileParser;
     switch (FileParser.ParseFile (pQFile->GetPointer())) {
     case iErr_File_OpenFailed:
@@ -404,8 +404,8 @@
     CDynamicString PollStr;
     FaxFormatType  format     = FaxFormat_Hylafax;
     tUInt          Resolution = 0;
-    char          *deviceName = DEFAULT_DEVICE_NAME;
-    char          *configName = 0;
+    char           deviceName[255] = DEFAULT_DEVICE_NAME;
+    char           configName[255] = DEFAULT_CONFIG_FILE;
     int            ch;
 
     // find device name
@@ -428,7 +428,13 @@
             break;
         case 'C':
             if (optarg && *optarg) {
-                configName = optarg;
+                //configName = optarg;
+                if (strlen(optarg) > sizeof(configName)-1) {
+                    printf("configName too long.");
+                    return 13;
+                }
+                
+                strncpy(configName, optarg, sizeof(configName)-1);
             }
             break;
         case 'd':
@@ -438,7 +444,12 @@
             break;
         case 'm':
             if (optarg && *optarg) {
-                deviceName = optarg;
+                if (strlen(optarg) > sizeof(deviceName)-1) {
+                    printf("deviceName too long.");
+                    return 13;
+                }
+                //deviceName = optarg;
+                strncpy(deviceName, optarg, sizeof(deviceName)-1);
             }
             break;
         case 'P':
@@ -466,9 +477,15 @@
             break;
         }
     }
-    if (!configName) {
-        configName = DEFAULT_CONFIG_FILE;
-    }
+    /*if (strlen(configName) < 1) {
+        if (strlen(DEFAULT_CONFIG_FILE) > sizeof(configName)-1) {
+            printf("default configName too long.");
+            return 13;
+        }
+  
+        strncpy(configName, DEFAULT_CONFIG_FILE, sizeof(configName)-1);
+        //configName = DEFAULT_CONFIG_FILE;
+    }*/
     if ((optind >= argc) && (PollStr.IsEmpty() == vTrue)) {
         printf ("No fax file specified!\n");
         return 1;
--- capi4hylafax-01.03.00/src/faxsend/faxsend.cpp	2016-10-16 21:31:43.975938770 +0200
+++ capi4hylafax_01.03.00_vb/src/faxsend/faxsend.cpp	2016-10-15 23:27:44.222168839 +0200
@@ -276,7 +276,7 @@
         dassert (FaxFilesList->IsEmpty() == vFalse);
     }
 
-    char *ModeText;
+    const char *ModeText;
     switch (GetFormat()) {
     default: {
         ModeText = "Hylafax";
@@ -550,7 +550,7 @@
         dassert (!PollRcvdCmd.IsEmpty());
         CDynamicString executeCommand;
         executeCommand.Set (&PollRcvdCmd);
-        char *StateText = "";
+        const char *StateText = "";
 
         if (GetFormat() == FaxFormat_Hylafax) {
             switch (GetRecvState()) {
@@ -686,9 +686,9 @@
 /*===========================================================================*\
 \*===========================================================================*/
 
-void CFaxSend::WriteLog (tSInt priority, char *text, ...) {
+void CFaxSend::WriteLog (tSInt priority, const char *text, ...) {
     dhead ("CFaxSend::WriteLog", DCON_CFaxSend);
-    char   *statusText = 0;
+    const char *statusText = NULL;
     char    timebuf[MAX_STRING_SIZE];
     timeval tv;
     va_list params;
--- capi4hylafax-01.03.00/src/standard/DbgInfo.cpp	2016-10-16 21:31:43.975938770 +0200
+++ capi4hylafax_01.03.00_vb/src/standard/DbgInfo.cpp	2016-10-16 12:41:31.630178224 +0200
@@ -22,6 +22,7 @@
 #include "aStrPrt.h"
 #include "aString.h"
 #include "ExtFuncs.h"
+#include <stdlib.h>
 
 /*---------------------------------------------------------------------------*\
 \*---------------------------------------------------------------------------*/
@@ -47,7 +48,7 @@
 /*===========================================================================*\
 \*===========================================================================*/
 
-void dwriteErr (unsigned DL, char *format, ...) {
+void dwriteErr (unsigned DL, const char *format, ...) {
     if (DL >= Dbg_Level_Errors) {
         char string[dwrite_MAX_CHAR_LENGTH];
         va_list param;
@@ -60,7 +61,7 @@
     }
 }
 
-void dwriteWrn (unsigned DL, char *format, ...) {
+void dwriteWrn (unsigned DL, const char *format, ...) {
     if (DL >= Dbg_Level_Infos) {
         char string[dwrite_MAX_CHAR_LENGTH];
         va_list param;
@@ -73,7 +74,7 @@
     }
 }
 
-void dwriteLog (unsigned DL, char *format, ...) {
+void dwriteLog (unsigned DL, const char *format, ...) {
     if (DL >= Dbg_Level_Logs) {
         char string[dwrite_MAX_CHAR_LENGTH];
         va_list param;
@@ -86,7 +87,7 @@
     }
 }
 
-void dwriteS (unsigned DL, char *format, ...) {
+void dwriteS (unsigned DL, const char *format, ...) {
     if (DL >= Dbg_Level_Short) {
         char string[dwrite_MAX_CHAR_LENGTH];
         va_list param;
@@ -99,7 +100,7 @@
     }
 }
 
-void dwriteE (unsigned DL, char *format, ...) {
+void dwriteE (unsigned DL, const char *format, ...) {
     if (DL >= Dbg_Level_Entry) {
         char string[dwrite_MAX_CHAR_LENGTH];
         va_list param;
@@ -112,7 +113,7 @@
     }
 }
 
-void dwriteI (unsigned DL, char *format, ...) {
+void dwriteI (unsigned DL, const char *format, ...) {
     if (DL >= Dbg_Level_Internal) {
         char string[dwrite_MAX_CHAR_LENGTH];
         va_list param;
@@ -238,22 +239,38 @@
 /*---------------------------------------------------------------------------*\
 \*---------------------------------------------------------------------------*/
 
-CDebugInfo::CDebugInfo(char *FuncName, unsigned DebugLevel, unsigned PrintLevel,
-                       char *_FileName, unsigned _LineNumber) {
+CDebugInfo::CDebugInfo(const char *FuncName, unsigned DebugLevel, unsigned PrintLevel,
+                       const char *_FileName, unsigned _LineNumber) {
     RetValue.format = 0;
     debuglevel = DebugLevel;
+    filename = NULL;
+    
     if (DebugLevel > Dbg_Level_Nothing) {
         spaces        = IndentSpaces;
         IndentSpaces += DEBUG_INDENT_WIDTH;
         name          = FuncName;
         linenumber    = _LineNumber;
         printlevel    = PrintLevel;
-        filename      = _FileName;
-        for (; *_FileName != '\0'; _FileName++) {
-            if ((*_FileName == '\\') || (*_FileName == '/')) {
-                filename = _FileName + 1;
+        //filename      = _FileName;
+        
+        unsigned int filenameLength = 0;
+        unsigned int pathLength = 0;
+        const char * curFileNameChar = _FileName;
+        
+        for (; *curFileNameChar != '\0'; curFileNameChar++) {
+            if ((*curFileNameChar == '\\') || (*curFileNameChar == '/')) {
+                //filename = _FileName + 1;
+                filenameLength = 0;
+            } else {
+                ++filenameLength;
             }
+            ++pathLength;
         }
+        
+        filename = (char *) malloc(filenameLength + 1);
+        s_strcpy(filename, _FileName+pathLength-filenameLength);
+    
+        
         if (printlevel & Prt_Level_NameAtFuncStart) {
             if (debuglevel == Dbg_Level_Short) {
                 RememberOutput (NAMEFORSHORT_BEGIN "%s%s", DbgIPrintLinePrefix, name);
@@ -271,7 +288,7 @@
         }
     } else {
         name       = 0;
-        filename   = 0;
+        //filename   = 0;
         linenumber = 0;
         printlevel = 0;
         spaces     = 0;
@@ -308,7 +325,8 @@
 
     /*----- Speicherplatz aufräumen -----*/
     name = 0;
-    filename = 0;
+    //filename = 0;
+    free(filename);
     linenumber = 0;
     debuglevel = printlevel = 0;
     spaces = 0;
--- capi4hylafax-01.03.00/src/standard/CString.h	2016-10-16 21:31:43.975938770 +0200
+++ capi4hylafax_01.03.00_vb/src/standard/CString.h	2016-10-16 20:58:10.964908255 +0200
@@ -43,14 +43,14 @@
 
 class CConstString {
 public:
-    CConstString (tString string, tSize len = 0);   // string have to be null terminated! len ist not tested!
+    CConstString (const tString string, tSize len = 0);   // string have to be null terminated! len ist not tested!
     virtual ~CConstString (void) {}
 
-    tBool   IsEmpty (void);
-    tSize   GetLen (void);      // GetLen & Get(Max)Size return multiples of tStringChar!
-    tSize   GetSize (void);
-    tSize   GetMaxSize (void);
-    tString GetPointer (void);
+    tBool   IsEmpty (void) const;
+    tSize   GetLen (void) const;      // GetLen & Get(Max)Size return multiples of tStringChar!
+    tSize   GetSize (void) const;
+    tSize   GetMaxSize (void) const;
+    tString GetPointer (void) const;
     tBool   ErrorDetected (void);
 
 protected:
@@ -69,41 +69,41 @@
 class CDynamicString : public CConstString {
 public:
     CDynamicString (void) {}
-    CDynamicString (tFormatChar *string, tSize len = 0);
-    CDynamicString (CConstString *pcstr);
+    CDynamicString (const tFormatChar *string, tSize len = 0);
+    CDynamicString (const CConstString *pcstr);
     CDynamicString (tUInt startSize);
     virtual ~CDynamicString (void);
 
-    tBool Set (tString       string, tSize len = 0);
-    tBool Set (tFormatChar  *string, tSize len = 0);
-    tBool Set (CConstString *pcstr);
-    tBool Append (tString       string, tSize len = 0);
-    tBool Append (tFormatChar  *string, tSize len = 0);
-    tBool Append (CConstString *pcstr);
-
-    tBool Change (tString       string, tSize startPos, tSize len = 0);
-    tBool Change (tFormatChar  *string, tSize startPos, tSize len = 0);
-    tBool Change (CConstString *pcstr,  tSize startPos, tSize len = 0);
-    tBool Insert (tString       string, tSize startPos, tSize len = 0);
-    tBool Insert (tFormatChar  *string, tSize startPos, tSize len = 0);
-    tBool Insert (CConstString *pcstr,  tSize startPos, tSize len = 0);
+    tBool Set (const tString       string, tSize len = 0);
+    tBool Set (const tFormatChar  *string, tSize len = 0);
+    tBool Set (const CConstString *pcstr);
+    tBool Append (const tString       string, tSize len = 0);
+    tBool Append (const tFormatChar  *string, tSize len = 0);
+    tBool Append (const CConstString *pcstr);
+
+    tBool Change (const tString       string, tSize startPos, tSize len = 0);
+    tBool Change (const tFormatChar  *string, tSize startPos, tSize len = 0);
+    tBool Change (const CConstString *pcstr,  tSize startPos, tSize len = 0);
+    tBool Insert (const tString       string, tSize startPos, tSize len = 0);
+    tBool Insert (const tFormatChar  *string, tSize startPos, tSize len = 0);
+    tBool Insert (const CConstString *pcstr,  tSize startPos, tSize len = 0);
 
     tBool Fill (tStringChar fillChar, tSize len, tSize startPos = 0);
     tBool FillAppend (tStringChar fillChar, tUInt len);
 
-    tSInt Print  (tFormatChar *format, ...);
-    tSInt vPrint (tFormatChar *format, va_list argptr);
-    tSInt PrintAppend  (tFormatChar *format, ...);
-    tSInt vPrintAppend (tFormatChar *format, va_list argptr);
+    tSInt Print  (const tFormatChar *format, ...);
+    tSInt vPrint (const tFormatChar *format, va_list argptr);
+    tSInt PrintAppend  (const tFormatChar *format, ...);
+    tSInt vPrintAppend (const tFormatChar *format, va_list argptr);
 
-    tCompare Compare (tString string, tSize len = 0);
-    tBool IsIdentical (tString string, tSize len = 0);
+    tCompare Compare (const tString string, tSize len = 0);
+    tBool IsIdentical (const tString string, tSize len = 0);
 
     void Get (tString string, tSize len = 0, tSize startPos = 0);
     tStringChar GetChar (tSize pos);
     tStringChar GetLastChar (void);
-    tSize FindChar (tStringChar ch, tSize startpos = 0); // not found => return = MAXVAL_tSize
-    tSize FindLastChar (tStringChar ch);                 // not found => return = MAXVAL_tSize
+    tSize FindChar (const tStringChar ch, tSize startpos = 0); // not found => return = MAXVAL_tSize
+    tSize FindLastChar (const tStringChar ch);                 // not found => return = MAXVAL_tSize
 
     void  SetLen (tSize newLen);
     void  CutLen (tSize cutLen);   // Result: newlen = oldLen - cutLen
@@ -113,7 +113,7 @@
     tBool Resize (tSize newMaxSize);
 
 private:
-    tSInt i_vPrintAppend (tFormatChar *format, va_list argptr, tUInt tabsize, tStringChar tabchar);
+    tSInt i_vPrintAppend (const tFormatChar *format, va_list argptr, tUInt tabsize, tStringChar tabchar);
     tUInt i_Print_ltoa (tFormatChar **pBuffer, tUInt64 value, tSByte radix, tUInt *pWidth, tUInt pFlags);
     tBool DynExpand (tSize newMaxSize);
 };
@@ -130,23 +130,23 @@
     detectError (vFalse) {
 }
 
-inline tBool CConstString::IsEmpty (void) {
+inline tBool CConstString::IsEmpty (void) const {
     return (curLen == 0);
 }
 
-inline tSize CConstString::GetLen (void) {
+inline tSize CConstString::GetLen (void) const {
     return curLen;
 }
 
-inline tSize CConstString::GetSize (void) {
+inline tSize CConstString::GetSize (void) const {
     return (GetLen()) ? curLen + 1 : 0;
 }
 
-inline tSize CConstString::GetMaxSize (void) {
+inline tSize CConstString::GetMaxSize (void) const {
     return maxSize;
 }
 
-inline tString CConstString::GetPointer (void) {
+inline tString CConstString::GetPointer (void) const {
     return (GetLen()) ? pntr : 0;
 }
 
@@ -163,17 +163,17 @@
     inline fuctions from CDynamicString
 \*===========================================================================*/
 
-inline tBool CDynamicString::Set (tString string, tSize len) {
+inline tBool CDynamicString::Set (const tString string, tSize len) {
     RemoveAll();
     return Append (string, len);
 }
 
-inline tBool CDynamicString::Set (tFormatChar *string, tSize len) {
+inline tBool CDynamicString::Set (const tFormatChar *string, tSize len) {
     RemoveAll();
     return Append (string, len);
 }
 
-inline tBool CDynamicString::Set (CConstString *pcstr) {
+inline tBool CDynamicString::Set (const CConstString *pcstr) {
     if (pcstr->GetLen()) {
         return Set (pcstr->GetPointer(), pcstr->GetLen());
     }
@@ -181,37 +181,37 @@
     return vTrue;
 }
 
-inline tBool CDynamicString::Append (tString string, tSize len) {
+inline tBool CDynamicString::Append (const tString string, tSize len) {
     return Change (string, GetLen(), len);
 }
 
-inline tBool CDynamicString::Append (tFormatChar *string, tSize len) {
+inline tBool CDynamicString::Append (const tFormatChar *string, tSize len) {
     return Change (string, GetLen(), len);
 }
 
-inline tBool CDynamicString::Append (CConstString *pcstr) {
+inline tBool CDynamicString::Append (const CConstString *pcstr) {
     return (pcstr->GetLen()) ? Change (pcstr->GetPointer(), GetLen(), pcstr->GetLen()) : vTrue;
 }
 
-inline tBool CDynamicString::Change (CConstString *pcstr, tSize startPos, tSize len) {
+inline tBool CDynamicString::Change (const CConstString *pcstr, tSize startPos, tSize len) {
     return (pcstr->GetLen()) ? Change (pcstr->GetPointer(), startPos, (len) ? len : pcstr->GetLen()) : vTrue;
 }
 
-inline tBool CDynamicString::Insert (CConstString *pcstr, tSize startPos, tSize len) {
+inline tBool CDynamicString::Insert (const CConstString *pcstr, tSize startPos, tSize len) {
     return (pcstr->GetLen()) ? Insert (pcstr->GetPointer(), startPos, (len) ? len : pcstr->GetLen()) : vTrue;
 }
 
 /*---------------------------------------------------------------------------*\
 \*---------------------------------------------------------------------------*/
 
-inline tBool CDynamicString::FillAppend (tStringChar fillChar, tUInt len) {
+inline tBool CDynamicString::FillAppend (const tStringChar fillChar, tUInt len) {
     return Fill (fillChar, len, GetLen());
 }
 
 /*---------------------------------------------------------------------------*\
 \*---------------------------------------------------------------------------*/
 
-inline tSInt CDynamicString::Print (tFormatChar *format, ...) {
+inline tSInt CDynamicString::Print (const tFormatChar *format, ...) {
     va_list arg_ptr;
     va_start (arg_ptr, format);
     tUInt fret = vPrint (format, arg_ptr);
@@ -219,12 +219,12 @@
     return fret;
 }
 
-inline tSInt CDynamicString::vPrint (tFormatChar *format, va_list argptr) {
+inline tSInt CDynamicString::vPrint (const tFormatChar *format, va_list argptr) {
     RemoveAll();
     return vPrintAppend (format, argptr);
 }
 
-inline tSInt CDynamicString::PrintAppend (tFormatChar *format, ...) {
+inline tSInt CDynamicString::PrintAppend (const tFormatChar *format, ...) {
     va_list arg_ptr;
     va_start (arg_ptr, format);
     tUInt fret = vPrintAppend (format, arg_ptr);
@@ -232,7 +232,7 @@
     return fret;
 }
 
-inline tSInt CDynamicString::vPrintAppend (tFormatChar *format, va_list argptr) {
+inline tSInt CDynamicString::vPrintAppend (const tFormatChar *format, va_list argptr) {
     return i_vPrintAppend (format, argptr, 0, 0);
 }
 
@@ -260,7 +260,13 @@
 inline tStringChar CDynamicString::GetLastChar (void) {
     //assert (GetLen() > 0);
     //assert (pntr != 0);
-    return pntr[GetLen() - 1];
+    int len = GetLen();
+    
+    if (len > 0) {
+        return pntr[GetLen() - 1];
+    } else {
+        return '\0';
+    }
 }
 
 inline void CDynamicString::SetLen (tSize newLen) {
--- capi4hylafax-01.03.00/src/standard/DbgInfo.h	2016-10-16 21:31:43.975938770 +0200
+++ capi4hylafax_01.03.00_vb/src/standard/DbgInfo.h	2016-10-16 21:27:43.802276702 +0200
@@ -86,12 +86,12 @@
 
 #else
 
-void dwriteErr (unsigned DL, char *format, ...);      // Logs
-void dwriteWrn (unsigned DL, char *format, ...);      // Logs
-void dwriteLog (unsigned DL, char *format, ...);      // Logs
-void dwriteS   (unsigned DL, char *format, ...);      // Short
-void dwriteE   (unsigned DL, char *format, ...);      // Entry
-void dwriteI   (unsigned DL, char *format, ...);      // Internal
+void dwriteErr (unsigned DL, const char *format, ...);      // Logs
+void dwriteWrn (unsigned DL, const char *format, ...);      // Logs
+void dwriteLog (unsigned DL, const char *format, ...);      // Logs
+void dwriteS   (unsigned DL, const char *format, ...);      // Short
+void dwriteE   (unsigned DL, const char *format, ...);      // Entry
+void dwriteI   (unsigned DL, const char *format, ...);      // Internal
 
 #endif
 
@@ -135,17 +135,17 @@
     TODO: noch nicht vollständig implementiert!!!!!
 \*---------------------------------------------------------------------------*/
 
-#define dhead(name, DL)             char *_CDI_Object_Name  = (name);               \
+#define dhead(name, DL)             const char *_CDI_Object_Name  = (const char *)(name);               \
                                     tUInt _CDI_Object_Level = (DL);                 \
-                                    if (_CDI_Object_Level >= Dbg_Level_Short) {     \
+                                    if (_CDI_Object_Level => Dbg_Level_Short) {     \
                                         DbgIPrintFull (NAMEATFUNCSTART_BEGIN "%s%s" \
                                                        NAMEATFUNCSTART_END,         \
                                                        DbgIPrintLinePrefix,         \
                                                        _CDI_Object_Name);           \
                                     }
-#define dheadP(name, DL, PL)        char *_CDI_Object_Name = (name);                \
+#define dheadP(name, DL, PL)        const char *_CDI_Object_Name = (const char *)(name);                \
                                     tUInt _CDI_Object_Level = (DL);                 \
-                                    if (_CDI_Object_Level >= Dbg_Level_Short) {     \
+                                    if (_CDI_Object_Level => Dbg_Level_Short) {     \
                                         DbgIPrintFull (NAMEATFUNCSTART_BEGIN "%s%s" \
                                                        NAMEATFUNCSTART_END,         \
                                                        DbgIPrintLinePrefix,         \
@@ -193,8 +193,8 @@
     Debug Makros & Funktionen
 \*---------------------------------------------------------------------------*/
 
-#define dhead(name, DL)             CDebugInfo _CDI_Object (name, DL, Prt_Level_Default, __FILE__, __LINE__)
-#define dheadP(name, DL, PL)        CDebugInfo _CDI_Object (name, DL, PL, __FILE__, __LINE__)
+#define dhead(name, DL)             CDebugInfo _CDI_Object ((const char *)name, DL, Prt_Level_Default, __FILE__, __LINE__)
+#define dheadP(name, DL, PL)        CDebugInfo _CDI_Object ((const char *)name, DL, PL, __FILE__, __LINE__)
 
 
 //void dparams (const char *format, ...);
@@ -236,8 +236,8 @@
 
 class CDebugInfo {
 public:
-    CDebugInfo (char *FuncName, unsigned DebugLevel, unsigned PrintLevel,
-                char *_FileName = "?", unsigned _LineNumber = 0);
+    CDebugInfo (const char *FuncName, unsigned DebugLevel, unsigned PrintLevel,
+                const char *_FileName = "?", unsigned _LineNumber = 0);
     virtual ~CDebugInfo (void);
 
     void print (const char *format, ...);
@@ -265,7 +265,7 @@
     void vRememberOutput (const char *format, va_list param);
 
 private:
-    char *name;
+    const char *name;
     char *filename;
     unsigned linenumber;
     unsigned debuglevel;
--- capi4hylafax-01.03.00/src/standard/PathName.h	2016-10-16 21:31:43.975938770 +0200
+++ capi4hylafax_01.03.00_vb/src/standard/PathName.h	2016-10-15 23:27:44.222168839 +0200
@@ -31,18 +31,18 @@
 class CPathName : public CDynamicString {
 public:
     CPathName (void);
-    CPathName (tFormatChar *GlobalPathName);
+    CPathName (const tFormatChar *GlobalPathName);
     CPathName (CDynamicString *GlobalPathName);
 
     tBool SetGlobal (CDynamicString *PathName);
-    tBool SetGlobal (tFormatChar *PathName, tSize len = 0);
+    tBool SetGlobal (const tFormatChar *PathName, tSize len = 0);
     tBool SetRelative (CDynamicString *RelPathName);
-    tBool SetRelative (tFormatChar *RelPathName, tSize len = 0);
+    tBool SetRelative (const tFormatChar *RelPathName, tSize len = 0);
 
     tUInt UpperPath (tUInt levels = 1);
     // return: the remaining levels (=> 0 always means success)
 
-    void SetDelimiter (tChar delimiter);
+    void SetDelimiter (const tChar delimiter);
 
 private:
     tChar m_Delimiter;
@@ -57,7 +57,7 @@
   : m_Delimiter (CPATHNAME_DELIMITER) {
 }
 
-inline CPathName::CPathName (tFormatChar *GlobalPathName)
+inline CPathName::CPathName (const tFormatChar *GlobalPathName)
   : CDynamicString (GlobalPathName),
     m_Delimiter (CPATHNAME_DELIMITER) {
 }
@@ -71,7 +71,7 @@
     return Set (PathName);
 }
 
-inline tBool CPathName::SetGlobal (tFormatChar *PathName, tSize len) {
+inline tBool CPathName::SetGlobal (const tFormatChar *PathName, tSize len) {
     return Set (PathName, len);
 }
 
@@ -91,7 +91,7 @@
     return Append (RelPathName);
 }
 
-inline tBool CPathName::SetRelative (tFormatChar *RelPathName, tSize len) {
+inline tBool CPathName::SetRelative (const tFormatChar *RelPathName, tSize len) {
     //dassert (RelPathName != 0);
     if (IsEmpty() == vFalse) {
         if (GetLastChar() == m_Delimiter) {
@@ -118,7 +118,7 @@
     return levels;
 }
 
-inline void CPathName::SetDelimiter (tChar delimiter) {
+inline void CPathName::SetDelimiter (const tChar delimiter) {
     m_Delimiter = delimiter;
 }
 
--- capi4hylafax-01.03.00/src/standard/aStrPrt.cpp	2016-10-16 21:31:43.979938666 +0200
+++ capi4hylafax_01.03.00_vb/src/standard/aStrPrt.cpp	2016-10-16 20:29:06.326394165 +0200
@@ -127,11 +127,13 @@
     tSInt PrtPreci   = -2;          // Precision of "placeholder"
     tUInt PrtValue   = 0;           // Returnvalue of convert function
     const char *savepos    = format;
-    char *argchar    = 0;
+    char  argchar[255]    = "";
     char  tabchar    = ' ';
     tUInt count      = maxcount;    // free place in buffer
     tUInt ErrorFound = 0;
 
+    char * va_arg_res = NULL;
+
     while ((count > 1) && (*format != 0) && (!ErrorFound)) {
         if (PrtFlags) {
 
@@ -196,10 +198,10 @@
             /*----- TYPE -----*/
             case 'b':
                 if ((char)va_arg (argptr, int)) {
-                    argchar = "TRUE";
+                    strncpy(argchar, "TRUE", sizeof(argchar));
                     PrtValue = 4;
                 } else {
-                    argchar = "FALSE";
+                    strncpy(argchar, "FALSE", sizeof(argchar));
                     PrtValue = 5;
                 }
                 if (PrtPreci == 1) {
@@ -282,10 +284,15 @@
                 if ((PrtFlags & PRINTFLAGS_QUESTION) && (va_arg (argptr, tUInt) > 1)) {
                     PrtFlags |= PRINTFLAGS_LARGE;
                 }
-                argchar = va_arg (argptr, char *);
-                if (!argchar) {
-                    PrtFlags &= ~PRINTFLAGS_LARGE;  // remove LARGE-FLAG
-                    argchar = "[NULL]";
+
+                va_arg_res = va_arg (argptr, char *);
+                
+                if (!va_arg_res) {
+                    PrtFlags &= ~PRINTFLAGS_LARGE;  // remove LARGE-FLAG    
+                    strncpy(argchar, "[NULL]", sizeof(argchar));
+                } else {
+                    strncpy(argchar, va_arg_res, sizeof(argchar));
+                    argchar[sizeof(argchar)-1] = '\0';
                 }
                 if (PrtFlags & PRINTFLAGS_LARGE) {
                     PrtValue = 0;
--- capi4hylafax-01.03.00/src/standard/ConfAssi.h	2016-10-16 21:31:43.979938666 +0200
+++ capi4hylafax_01.03.00_vb/src/standard/ConfAssi.h	2016-10-15 23:27:44.222168839 +0200
@@ -31,7 +31,7 @@
 };
 
 struct ConfigAssignment_Item {
-    char                  *Name;
+    const char            *Name;
     ConfigAssignment_Types Type;
     void                  *Data;
 };
--- capi4hylafax-01.03.00/src/standard/ExtFuncs.h	2016-10-16 21:31:43.979938666 +0200
+++ capi4hylafax_01.03.00_vb/src/standard/ExtFuncs.h	2016-10-16 21:10:09.062162548 +0200
@@ -24,7 +24,7 @@
 
 #ifndef NDEBUG
 
-    #define USE_DEBUG_INFO_CLASS        2
+    #define USE_DEBUG_INFO_CLASS        7
     #define DEBUG_INDENT_WIDTH          0
     //#define DEBUG_FILE_NAME             "/var/log/c2faxfcalls.log"
 
@@ -63,10 +63,10 @@
 
 #define DbgIAssert(a)
 
-extern char    *DbgIPrintLinePrefix;
+extern const char *DbgIPrintLinePrefix;
 extern unsigned DbgIPrintLinePrefixLen;
 
-extern void DbgIPrintStr (char *output);
+extern void DbgIPrintStr (const char *output);
 
 extern void DbgIBreakPoint (void);
 
--- capi4hylafax-01.03.00/src/standard/CString.cpp	2016-10-16 21:31:43.979938666 +0200
+++ capi4hylafax_01.03.00_vb/src/standard/CString.cpp	2016-10-15 23:27:44.222168839 +0200
@@ -51,7 +51,7 @@
 /*===========================================================================*\
 \*===========================================================================*/
 
-CConstString::CConstString (tString string, tSize len)
+CConstString::CConstString (const tString string, tSize len)
   : pntr (string),
     detectError (vFalse) {
 
@@ -67,7 +67,7 @@
 /*===========================================================================*\
 \*===========================================================================*/
 
-CDynamicString::CDynamicString (tFormatChar *string, tSize len) {
+CDynamicString::CDynamicString (const tFormatChar *string, tSize len) {
     dhead ("CDynamicString-Constructor", DCON_CDynamicString);
     dparams ("%s,%x", string, len);
     Set (string, len);
@@ -76,7 +76,7 @@
 /*===========================================================================*\
 \*===========================================================================*/
 
-CDynamicString::CDynamicString (CConstString *pcstr) {
+CDynamicString::CDynamicString (const CConstString *pcstr) {
     dhead ("CDynamicString-Constructor", DCON_CDynamicString);
     dparams ("%s", pcstr);
     Set (pcstr);
@@ -108,7 +108,7 @@
 /*===========================================================================*\
 \*===========================================================================*/
 
-tBool CDynamicString::Change (tString string, tSize startPos, tSize len) {
+tBool CDynamicString::Change (const tString string, tSize startPos, tSize len) {
     dhead ("CDynamicString::Change", DCON_CDynamicString);
     dparams ("%s,%x,%x", string, startPos, len, pntr);
     dassert (string != 0);
@@ -136,7 +136,7 @@
 /*===========================================================================*\
 \*===========================================================================*/
 
-tBool CDynamicString::Change (tFormatChar *string, tSize startPos, tSize len) {
+tBool CDynamicString::Change (const tFormatChar *string, tSize startPos, tSize len) {
     dhead ("CDynamicString::Change", DCON_CDynamicString);
     dparams ("%s,%x,%x,%s", string, startPos, len, pntr);
     dassert (string != 0);
@@ -164,7 +164,7 @@
 /*===========================================================================*\
 \*===========================================================================*/
 
-tBool CDynamicString::Insert (tString string, tSize startPos, tSize len) {
+tBool CDynamicString::Insert (const tString string, tSize startPos, tSize len) {
     dhead ("CDynamicString::Insert", DCON_CDynamicString);
     dassert (string != 0);
     dassert (startPos <= GetLen());
@@ -191,7 +191,7 @@
 /*===========================================================================*\
 \*===========================================================================*/
 
-tBool CDynamicString::Fill (tStringChar fillChar, tSize len, tSize startPos) {
+tBool CDynamicString::Fill (const tStringChar fillChar, tSize len, tSize startPos) {
     dhead ("CDynamicString::Fill", DCON_CDynamicString);
     dparams ("%c,%x,%x,%s", fillChar, len, startPos, pntr);
     dassert (startPos <= GetLen());
@@ -350,11 +350,11 @@
 /*===========================================================================*\
 \*===========================================================================*/
 
-tSInt CDynamicString::i_vPrintAppend (tFormatChar *format, va_list argptr, tUInt tabsize, tStringChar tabchar) {
+tSInt CDynamicString::i_vPrintAppend (const tFormatChar *format, va_list argptr, tUInt tabsize, tStringChar tabchar) {
     dhead ("CDynamicString::i_vPrintAppend", DCON_CDynamicString);
     dassert (format != 0);
     tFormatChar  localBuf[CSTRING_PRINT_BufferSize];    // Local internal buffer
-    tFormatChar *savepos    = format;                   // Position of last "placeholder"
+    const tFormatChar *savepos = format;                   // Position of last "placeholder"
     tFormatChar *copyBuf    = &localBuf[0];             // Buffer that will be append to String
     tSize        copyBufLen = 0;                        // Size of copyBuf
     tUInt        PrtFlags   = 0;                        // Possible Flags for "placeholder"
@@ -412,9 +412,9 @@
             case '8':
             case '9':
                 if (PrtPreci == -2) {
-                    PrtWidth = a_strtol (format, &format, 10);
+                    PrtWidth = a_strtol ((char*)format, (char**)&format, 10);
                 } else {
-                    PrtPreci = a_strtol (format, &format, 10);
+                    PrtPreci = a_strtol ((char*)format, (char**)&format, 10);
                 }
                 format--;
                 break;
@@ -570,8 +570,14 @@
 
             /*----- OTHERS: wrong placeholder -----*/
             default:
-                copyBuf    = savepos;
+                //copyBuf    = (char*)savepos; // todo: avoid this cast
+                copyBuf = &localBuf[0];
+                //localBuf[CSTRING_PRINT_BufferSize-1] = '\0';
+                strncpy(copyBuf, savepos, CSTRING_PRINT_BufferSize);
                 copyBufLen = format - savepos + 1;
+                if (copyBufLen > CSTRING_PRINT_BufferSize)
+                    copyBufLen = CSTRING_PRINT_BufferSize;
+                
                 PrtFlags  |= PRINTFLAGS_READY;
                 break;
             } //end: switch (*format)
@@ -662,7 +668,7 @@
 				++curLen;
 			    }
 			} else {
-			    dassert((sizeof tChar)== (sizeof tFormatChar));
+			    dassert((sizeof(tChar))== (sizeof(tFormatChar)));
 			    const tChar *p = (tChar*)copyBuf;
 			    const tChar * const p_end = p + copyBufLen;
 			    for (;p < p_end; ++p) {
--- capi4hylafax-01.03.00/src/linuxdep/dbgLEVEL.cpp	2016-10-16 21:31:43.979938666 +0200
+++ capi4hylafax_01.03.00_vb/src/linuxdep/dbgLEVEL.cpp	2016-10-15 23:27:44.226168733 +0200
@@ -72,14 +72,14 @@
 /*===========================================================================*\
 \*===========================================================================*/
 
-char    *DbgIPrintLinePrefix    = "";
+const char *DbgIPrintLinePrefix = "";
 unsigned DbgIPrintLinePrefixLen = 0;
 
 #if defined (DCON_FILE_NAME) && (USE_DCON_INFO_CLASS > 0)
 
 #include "DbgFile.h"
 
-void DbgIPrintStr (char *output) {
+void DbgIPrintStr (const char *output) {
     DebugFilePrint (output);
 }
 
@@ -87,7 +87,7 @@
 
 #include <stdio.h>
 
-void DbgIPrintStr (char *output) {
+void DbgIPrintStr (const char *output) {
     puts (output);
 }
 
--- capi4hylafax-01.03.00/src/linuxdep/misc.cpp	2016-10-16 21:31:43.979938666 +0200
+++ capi4hylafax_01.03.00_vb/src/linuxdep/misc.cpp	2016-10-15 23:27:44.226168733 +0200
@@ -110,7 +110,7 @@
 \*---------------------------------------------------------------------------*/
 
 const char *DataFormat2str (tUInt DataFormat) {
-    char *t;
+    const char *t;
     switch (DataFormat) {
     case 0:
         t = "SFF";
