Package:     capi4hylafax
Subject:     fix for run autoconf
Author:      Joachim Wiedorn <ad_debian at joonet.de>
Forwarded:   no
Last-Update: 2012-06-12

Fix warnings with newer automake tools.
---

diff -urN s03/configure.in s04/configure.in
--- s03/configure.in	2006-08-15 11:13:25.000000000 +0200
+++ s04/configure.in	2012-06-12 19:57:00.436291221 +0200
@@ -17,7 +17,7 @@
 AC_PROG_LIBTOOL
 
 dnl Check for va_copy in stdarg.h
-AC_CACHE_CHECK([for va_copy],ac_func_va_copy,[
+AC_CACHE_CHECK([for va_copy],ac_func_cv_va_copy,[
 AC_LINK_IFELSE([AC_LANG_PROGRAM([
 #include <stdarg.h>
 void x(va_list args)
@@ -25,11 +25,11 @@
     va_list copy;
     va_copy(copy,args);
 }
-], []) ], [ac_func_va_copy=yes], [ac_func_va_copy=no])
+], []) ], [ac_func_cv_va_copy=yes], [ac_func_cv_va_copy=no])
 ])
 
-if test $ac_func_va_copy = no; then
-    AC_CACHE_CHECK([for __va_copy],ac_func___va_copy,[
+if test $ac_func_cv_va_copy = no; then
+    AC_CACHE_CHECK([for __va_copy],ac_func_cv___va_copy,[
         AC_LINK_IFELSE([AC_LANG_PROGRAM([
         #include <stdarg.h>
         void x(va_list args)
@@ -37,9 +37,9 @@
            va_list copy;
            __va_copy(copy,args);
         }
-    ], [])], [ac_func___va_copy=yes], [ac_func___va_copy=no])
+    ], [])], [ac_func_cv___va_copy=yes], [ac_func_cv___va_copy=no])
     ])
-    if test $ac_func___va_copy = no; then
+    if test $ac_func_cv___va_copy = no; then
        AC_DEFINE(va_copy(d,s),(d)=(s))
     else
        AC_DEFINE(va_copy(d,s),__va_copy(d,s))
